coderd/unhanger/detector.go

```diff
@@ -0,0 +1,332 @@
+	ctx, cancel := context.WithTimeout(d.ctx, 5*time.Minute)
+	defer cancel()
+
+	stats := Stats{
+		HungJobIDs: []uuid.UUID{},
+		Error:      nil,
+	}
+
+	err := d.db.InTx(func(db database.Store) error {
+		locked, err := db.TryAcquireLock(ctx, database.LockIDHangDetector)
+		if !locked {
+			// If we can't acquire the lock, it means another instance of the
+			// hang detector is already running in another coder replica.
+			// There's no point in waiting to run it again, so we'll just retry
+			// on the next tick.
+			d.log.Info(ctx, "skipping workspace build hang detector run due to lock", slog.Error(err))
+			// This error is ignored.
+			return &acquireLockError{err: err}
+		}
+		if err != nil {
+			d.log.Warn(ctx, "skipping workspace build hang detector run due to error acquiring lock", slog.Error(err))
+			return xerrors.Errorf("acquire lock: %w", err)
+		}
+		d.log.Info(ctx, "running workspace build hang detector")
+
+		// Find all provisioner jobs that are currently running but have not
+		// received an update in the last 5 minutes.
+		jobs, err := db.GetHungProvisionerJobs(ctx, t.Add(-HungJobDuration))
+		if err != nil {
+			return xerrors.Errorf("get hung provisioner jobs: %w", err)
+		}
+
+		// Send a message into the build log for each hung job saying that it
+		// has been detected and will be terminated, then mark the job as
+		// failed.
+		for _, job := range jobs {
+			log := d.log.With(slog.F("job_id", job.ID))
+
+			jobStatus := db2sdk.ProvisionerJobStatus(job)
+			if jobStatus != codersdk.ProvisionerJobRunning {
+				log.Error(ctx, "hang detector query discovered non-running job, this is a bug", slog.F("status", jobStatus))
+				continue
+			}
+
+			log.Info(ctx, "detected hung (>5m) provisioner job, forcefully terminating")
+
+			// First, get the latest logs from the build so we can make sure
+			// our messages are in the latest stage.
+			logs, err := db.GetProvisionerLogsAfterID(ctx, database.GetProvisionerLogsAfterIDParams{
+				JobID:        job.ID,
+				CreatedAfter: 0,
+			})
+			if err != nil {
+				log.Warn(ctx, "get logs for hung job", slog.Error(err))
+				continue
+			}
+			logStage := ""
+			if len(logs) != 0 {
+				logStage = logs[len(logs)-1].Stage
+			}
+			if logStage == "" {
+				logStage = "Unknown"
+			}
+
+			// Insert the messages into the build log.
+			insertParams := database.InsertProvisionerJobLogsParams{
+				JobID: job.ID,
+			}
+			now := database.Now()
+			for i, msg := range HungJobLogMessages {
+				// Set the created at in a way that ensures each message has
+				// a unique timestamp so they will be sorted correctly.
+				insertParams.CreatedAt = append(insertParams.CreatedAt, now.Add(time.Millisecond*time.Duration(i)))
+				insertParams.Level = append(insertParams.Level, database.LogLevelError)
+				insertParams.Stage = append(insertParams.Stage, logStage)
+				insertParams.Source = append(insertParams.Source, database.LogSourceProvisionerDaemon)
```

Comment:
We don't display where the logs came from anywhere ATM, if we wanted to do that at some point this could be worth it but currently this enum doesn't really do anything AFAICT